### PR TITLE
feat: make generator use template files

### PR DESCRIPTION
**Pull request by dongchangYoo**
Existing `code-generators` create files line by line as strings, making it difficult to see at a glance what file will be generated. This PR solves the problem by having the `code-generator` substitute template files to create each file.

### CODE REVIEW

File: tachyon/math/elliptic_curves/bn/generator/curve.h.tpl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  constexpr static TwistType kTwistType = TwistType::k%{twist_type};
+
+  using Fp = Fq;
+  using Fp2 = Fq2;
+  using Fp6 = Fq6;
+  using Fp12 = Fq12;
+  using G1Curve = _G1Curve;
+  using G2Curve = _G2Curve;
+
+  // NOTE(chokobole): Make them constexpr."
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/bn/generator/curve.h.tpl:31]:
```suggestion
  // NOTE(chokobole): Make them constexpr.
```

File: tachyon/math/elliptic_curves/bn/generator/curve.h.tpl [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// clang-format off
+#include "tachyon/base/logging.h"
+#include "tachyon/math/elliptic_curves/bn/bn_curve.h"
+#include "%{fq12_hdr}"
+#include "%{g1_hdr}"
+#include "%{g2_hdr}"
+#include "tachyon/math/elliptic_curves/pairing/twist_type.h"
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/bn/generator/curve.h.tpl:7]:
```suggestion
#include "tachyon/math/elliptic_curves/bn/bn_curve.h"
#include "tachyon/math/elliptic_curves/pairing/twist_type.h"
#include "%{fq12_hdr}"
#include "%{g1_hdr}"
#include "%{g2_hdr}"
```

File: tachyon/math/elliptic_curves/short_weierstrass/generator/generator.cc [hunk 3]
```diff
@@ -11,7 +11,20 @@
...
 namespace tachyon {
 
+std::string GenerateMulByAFunc(bool mul_by_a_fast, int64_t a_value) {
+  if (!mul_by_a_fast) return "    return kA * v;";
+  if (a_value == 0) return "    return BaseField::Zero();";
+
+  std::stringstream ss;
+  ss << "    return ";
+  ss << math::GenerateFastMultiplication(a_value) << ";";
+  return ss.str();
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/short_weierstrass/generator/generator.cc:21]:
```suggestion
  if (a_value == 0) return "    return BaseField::Zero();";
  if (mul_by_a_fast) {
    std::stringstream ss;
    ss << "    return ";
    ss << math::GenerateFastMultiplication(a_value) << ";";
    return ss.str();
  }
  return "    return kA * v;";
```

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc [hunk 4]
```diff
@@ -1,6 +1,10 @@
+#include <map>
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc:1]: Please remove this!

File: tachyon/math/finite_fields/generator/ext_prime_field_generator/ext_prime_field_generator.cc [hunk 5]
```diff
@@ -1,3 +1,6 @@
+#include <map>
```
Comment (chokobole) [inline tachyon/math/finite_fields/generator/ext_prime_field_generator/ext_prime_field_generator.cc:1]: Please remove this

File: tachyon/math/elliptic_curves/bn/generator/generator.cc [hunk 6]
```diff
@@ -1,3 +1,5 @@
+#include <map>
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/bn/generator/generator.cc:1]: Please remove this

File: tachyon/math/elliptic_curves/bls12/generator/curve.h.tpl [hunk 7]
```diff
@@ -0,0 +1,38 @@
+// clang-format off
+#include "tachyon/base/logging.h"
+#include "tachyon/math/elliptic_curves/bls12/bls12_curve.h"
+#include "%{fq12_hdr}"
+#include "%{g1_hdr}"
+#include "%{g2_hdr}"
+#include "tachyon/math/elliptic_curves/pairing/twist_type.h"
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/bls12/generator/curve.h.tpl:7]:
```suggestion
#include "tachyon/math/elliptic_curves/bls12/bls12_curve.h"
#include "tachyon/math/elliptic_curves/pairing/twist_type.h"
#include "%{fq12_hdr}"
#include "%{g1_hdr}"
#include "%{g2_hdr}"
```

**Review by chokobole — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/build/cc_writer.h [hunk 8]
```diff
@@ -14,6 +15,21 @@ struct CcWriter : public Writer {
...
   int WriteSrc(const std::string& content) const;
+
+  // Remove all lines equals to "%{if |tag|}" or "%{endif |tag|}"
+  // if |select_tag_block| is true. Otherwise, remove all lines that are
+  // enclosed between "%{if |tag|}" and "%{endif |tag|}".
+  static void RemoveOptionalLines(std::vector<std::string>& tpl_lines,
+                                  std::string_view tag, bool select_tag_block);
+
+ private:
+  // Remove the each first line that equals to |start_line| and |end_line| if
```
Comment (ashjeong) [inline tachyon/build/cc_writer.h:26]:
```suggestion
  // Remove each first line that equals to |start_line| and |end_line| if

**Review by ashjeong — CHANGES_REQUESTED** [review]
Note - I also changed the PR's title.

**Review by ashjeong — APPROVED** [review]
LGTM

**Pull request merged**